### ISSUE TITLE
Getting the error when loading the .pb file using cv2.dnn.readNetFromTensorflow

### ISSUE THREAD

**Issue by srinivas1746** [opencv/opencv#18759]
##### System information (version)

\

##### Detailed description

<  My model has dsnt(Numerical Coordinate Regression with Convolutional Neural Networks) custom layer on top of keras functional model. I have generated .pb file using custom tensorflow script.When I remove the dsnt layer from the model then the graph is loading properly>

#####Error
<    net = cv2.dnn.readNetFromTensorflow(modelWeights)
cv2.error: OpenCV(4.4.0) /tmp/pip-req-build-6amqbhlx/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:1025: error: (-215:Assertion failed) values.type() == CV_32FC1 in function 'populateNet'
>

**Comment by dkurt** [comment]
Please try the latest OpenCV version (4.5.0). If it does not work - attach the model to make an issue reproducible.

**Comment by srinivas1746** [comment]
Thank you for your responce @dkurt .I have tried with opencv 4.5.0.But it is still giving the same error.Here I am attached the .pb file
[frozen_graph_latest_regression_dsnt.zip](https://github.com/opencv/opencv/files/5509053/frozen_graph_latest_regression_dsnt.zip)

**Comment by srinivas1746** [comment]
Meanwhile I have created the .pb file with different custom method.I am getting the below error

#### Error
<
opencv_4-5-0/opencv/modules/dnn/src/tensorflow/tf_graph_simplifier.cpp:904: error: (-2:Unspecified error) Tensor's data type is not supported in function 'getTensorContent'

>

**Comment by srinivas1746** [comment]
### .pb file generation code

<
sess = K.get_session()

# Write the graph in human readable
f = 'graph_def_for_reference.pb.ascii'
tf.compat.v1.train.write_graph(sess.graph.as_graph_def(), '.', f, as_text=True)

input_graph_def = sess.graph.as_graph_def()

#freeze graph
from tensorflow.python.framework.graph_util import convert_variables_to_constants
output_names = pred_node_names
output_names += [v.op.name for v in tf.compat.v1.global_variables()]
print(output_names)
constant_graph = convert_variables_to_constants(sess, input_graph_def,output_names)

# Write the graph in binary .pb file
from tensorflow.python.framework import graph_io
graph_io.write_graph(constant_graph, '.', 'reg_dsnt.pbtxt', as_text=True)
graph_io.write_graph(constant_graph, '.', 'reg_dsnt.pb', as_text=False)

>

**Comment by srinivas1746** [comment]
@dkurt  Here is the link for .pb file and tensorboard logs for graph visualization.
https://drive.google.com/drive/folders/1IxJ6kcqbf40guY8ukN8r5IKx6vFoHdA9?usp=sharing

**Comment by l-bat** [comment]
@srinivas1746 I looked at your model, it contains operations:` Range, Tile` which are not supported in OpenCV

**Comment by dkurt** [comment]
@srinivas1746, please shared model definition. Probably, these unsupported layers are generated by a single layer so we can fuse them.

**Comment by srinivas1746** [comment]
#### My dsnt model code is here

```
def dsnt(inputs, method='softmax'):
    '''
    Differentiable Spatial to Numerical Transform, as taken from the paper "Numerical Coordinate
    Regression with Convolutional Neural Networks"
    Arguments:
        inputs - The learnt heatmap. A 3d tensor of shape [batch, height, width]
        method - A string representing the normalisation method. See `_normalise_heatmap` for available methods
    Returns:
        norm_heatmap - The given heatmap with normalisation/rectification applied
        coords_zipped - A tensor of shape [batch, 2] containing the [x, y] coordinate pairs
    '''
    # Rectify and reshape inputs
    norm_heatmap = _normalise_heatmap(inputs, method)

    batch_count = tf.shape(norm_heatmap)[0]
    height = tf.shape(norm_heatmap)[1]
    width = tf.shape(norm_heatmap)[2]

    # Build the DSNT x, y matrices
    dsnt_x = tf.tile([[(2 * tf.range(1, width + 1) - (width + 1)) / width]], [batch_count, height, 1])
    dsnt_x = tf.cast(dsnt_x, tf.float32)
    dsnt_y = tf.tile([[(2 * tf.range(1, height + 1) - (height + 1)) / height]], [batch_count, width, 1])
    dsnt_y = tf.cast(tf.transpose(dsnt_y, perm=[0, 2, 1]), tf.float32)

    # Compute the Frobenius inner product
    outputs_x = tf.reduce_sum(tf.multiply(norm_heatmap, dsnt_x), axis=[1, 2])
    outputs_y = tf.reduce_sum(tf.multiply(norm_heatmap, dsnt_y), axis=[1, 2])

    # Zip into [x, y] pairs
    coords_zipped = tf.stack([outputs_x, outputs_y], axis=1)
    coords_zipped = (coords_zipped + 1) / 2

    return coords_zipped
def _normalise_heatmap(inputs, method='softmax'):
    '''
    Applies the chosen normalisation/rectification method to the input tensor
    Arguments:
        inputs - A 4d tensor of shape [batch, height, width, 1] (the learnt heatmap)
        method - A string representing the normalisation method. One of those shown below
    '''
    # Remove the final dimension as it's of size 1
    inputs = tf.reshape(inputs, tf.shape(inputs)[:3])

    # Normalise the values such that the values sum to one for each heatmap
    normalise = lambda x: tf.div(x, tf.reshape(tf.reduce_sum(x, [1, 2]), [-1, 1, 1]))

    # Perform rectification
    if method == 'softmax':
        inputs = _softmax2d(inputs, axes=[1, 2])
    elif method == 'abs':
        inputs = tf.abs(inputs)
        inputs = normalise(inputs)
    elif method == 'relu':
        inputs = tf.nn.relu(inputs)
        inputs = normalise(inputs)
    elif method == 'sigmoid':
        inputs = tf.nn.sigmoid(inputs)
        inputs = normalise(inputs)
    else:
        msg = "Unknown rectification method \"{}\"".format(method)
        raise ValueError(msg)
    return inputs
def _softmax2d(target, axes):
    '''
    A softmax implementation which can operate across more than one axis - as this isn't
    provided by Tensorflow
    Arguments:
        targets - The tensor on which to apply softmax
        axes - An integer or list of integers across which to apply softmax
    '''
    max_axis = tf.reduce_max(target, axes, keepdims=True)
    target_exp = tf.exp(target - max_axis)
    normalize = tf.reduce_sum(target_exp, axes, keepdims=True)
    softmax = target_exp / normalize
    return softmax
#### I am passing the heatmap(output of Maxpool) to the dsnt function
outputs=dsnt(output)

```

**Comment by rogday** [comment]
After a model diagnostic run, the following problems arose: `Shape`, `Cast`, `Range`, `Tile` layers aren't supported; `parseBias()` and `parseMul()` assert that constants are `32FC1`; `Pack` doesn't support constant inputs, as well as `axis` not being equal to 0.
